### PR TITLE
[WIP] Feature/remote repo

### PR DESCRIPTION
**Pull request by DavidPoliakoff**
Currently in a straw man state, I wouldn't yet merge this. The feature I want:

``` bash
spack repo add https://www.github.com/DavidPoliakoff/some_package_repository.git
```

To download the given github package and add it as a repo (this will be useful for site-local collections of spack packages)

Some questions remain:
- How do we get the fetch strategy from the url (in fetch_strategy.py, the function from_url, currently a straw man)?
- Where do we want these repositories to live? Right now I have them being stored in a package underneath the main repo, but could as easily see them living at the same level as the builtin collection of packages.

Thoughts? This is as much an issue with a strawman attached as a PR.

### CODE REVIEW

File: lib/spack/spack/cmd/repo.py [hunk 1]
```diff
@@ -83,7 +88,22 @@ def repo_create(args):
...
+            fetcher.fetch()
+            packageName = os.listdir(stage.path)[0]
+            try:
+                shutil.rmtree(root_repo.packages_path + "/" + packageName)
+            except OSError:
+                pass
+            shutil.move(
+                stage.path + "/" + packageName,
+                root_repo.packages_path)
+            path = root_repo.packages_path + "/" + packageName
```
Comment (DavidPoliakoff) [inline lib/spack/spack/cmd/repo.py]: All of these are written in Python-centric python, and should probably be in more of a spack idiom. Curious to know the correct way to say this.

**Comment by tgamblin** [comment]
marking WIP for now.

**Comment by drbenmorgan** [comment]
@DavidPoliakoff though this a WIP, I'm very interested in this feature 👍 (coming from a Homebrew/Taps background). I'm a spack newbie so comments below might reflect experience with homebrew/taps rather than how spack does things...

> How do we get the fetch strategy from the url (in fetch_strategy.py, the function from_url, currently a straw man)?

Could an extra argument be added to specify this, e.g. 

```
spack repo add --fetch=git https://www.github.com/DavidPoliakoff/some_package_repository.git
```

or failing that some kind of lookup mapping the protocol and/or host to strategy (e.g. if `https://www.github.com` it's almost certain that `git clone` is needed)?

> 
> Where do we want these repositories to live? Right now I have them being stored in a package underneath the main repo, but could as easily see them living at the same level as the builtin collection of packages.

My preference would be to put them at the same level as the spack builtin by default, e.g.

```
+- spack.git/
   +- var/
      +- spack/
         +- repos/
            +- builtin
            +- myrepo
            +- anotherrepo
```

that seems cleanest given adding a repo via a fetch feels like asking spack to take responsibility for where it goes. I guess one could also use a ``--directory`` argument to specify somewhere else like ``spack repo create``. There's also the site/user configuration scope which I'm not so familiar with yet.

Perhaps this is going beyond the scope of this PR, but if a repo is fetched from a VCS by spack, could/should spack have (directly or via a new subcommand of repo) a facility to update it? That also maybe applies to whether a ``spack update`` command is considered useful/required.

**Comment by DavidPoliakoff** [comment]
Cool, modified to put repo in that location. @tgamblin thoughts on a --fetch argument? Perhaps mirroring the package syntax, could use --git, --hg for trivial implementation, or we could truly implement from_url and avoid (command) arguments

**Comment by tgamblin** [comment]
@DavidPoliakoff: I'd rather not have `--fetch` argument, esp. when you could figure out the protocol from the URL.  If that means we only support git, or git and http[s], for doing this, then I'm not really going to be too sad about it...

**Comment by mwilliammyers** [comment]
Being able to explicitly give a git URL is awesome but coming from a Homebrew background (my previous attempt at an HPC focused package manager was hacking Linuxbrew), we might also want to consider a naming convention. Eg. `spack repo add DavidPoliakoff/some_package_repository` would look for the URL `https://www.github.com/DavidPoliakoff/some_package_repository.git` or `https://www.github.com/DavidPoliakoff/spack-some_package_repository.git` (or preferably it would look for both and be flexible).

**Comment by tgamblin** [comment]
@DavidPoliakoff: as of #1015 there is a `file_cache` module in Spack that lets you lock and manipulate storage in the user's home directory, in `~/.spack/cache`.  It's currently used to maintain a cache of virtual package providers so that we don't have to scan all the `package.py`'s all the time, but you could use it as a local store for remote repos.

**Comment by DavidPoliakoff** [comment]
Don't have time to do this, closing. Feel free to reopen if anybody else wants to push on it

**Pull request closed without merging**